doc/watermarks_HLD.md

```diff
@@ -0,0 +1,212 @@
+Interface                    MC0   MC1   MC2   MC3   MC4   MC5   MC6   MC7
+Ethernet0                      0     3     0     0     0     0     0     0
+…
+Ethernet128                    0     0     0     0     0     0     0     0
+```
+
+In addition clear functionality will be added:
+
+$ clear priority-group [watermark|persistent-watermark] headroom
+
+$ clear priority-group [watermark|persistent-watermark] shared
+
+$ clear queue [watermark|persistent-watermark] unicast
+
+$ clear queue [watermark|persistent-watermark] mutlicast
+
+### 3.1.2 Counter DB
+
+#### The following new Queue counters should be available for each queue entry in the DB:
+- "COUNTERS:queue_vid"
+  - SAI_QUEUE_STAT_SHARED_WATERMARK_BYTES
+#### For every Ingress PG the following should be available in the DB:
+- "COUNTERS:pg_vid"
+  - SAI_INGRESS_PRIORITY_GROUP_STAT_XOFF_ROOM_WATERMARK_BYTES
+  - SAI_INGRESS_PRIORITY_GROUP_STAT_SHARED_WATERMARK_BYTES
+#### Additionally a few mappings should be added:
+- "COUNTERS_PG_PORT_MAP" - map PG oid to port oid
+- "COUNTERS_PG_NAME_MAP" - map PG oid to PG name
+- "COUNTERS_PG_INDEX_MAP" - map PG oid to PG index
+
+#### New tables will be introduced: 
+ - "TELEMETRY_WATERMARKS" - containing the watermarks for the telemetry to consume
+ - "PERSISTENT_WATERMARKS" - containing persistent watermarks, cleared only by user and not by telemetry.
+
+### 3.1.3 Orchestration Agent
+Portorch should be updated:
+ - add SAI_QUEUE_STAT_SHARED_WATERMARK_BYTES to queueStatIds
+ - implement new counter group for PG and priorityGroupStatIds
+ - implement PG to port map generation
+
+New watermark orch should be implemented with the following functionality:
+ - Read TELEMETRY_PERIOD from config DB, start a timer for periodically providing telemetry watermark.
+ - Every telemetry period put current watermark into TELEMETRY_WATERMARK table, than clear it on hardware (with clear stat api).
+   If the watermark was cleared during last period by the user, compare current watermark with  value cached before user clear, and submit the highest one. 
+ - If received USER_CLEAR notification, save current watermark in cache for current telemetry interval, clear the watermark on hardware.
+ - If received USER_CLEAR_PERSISTENT notification, the PERSISTENT_WATERMARK gets overwriten by current watermark from counter DB.
+ - Current watermark periodically gets compared with persistant watermark and overwrites it if it's higher. Persistant watermark also gets updated before any clearing of watermarks on hardware.
+
+### 3.1.4 SAI Redis
+Flex counter should be extended to support new PG counters.
+### 3.1.5 CONFIG DB
+Add new table WATERMARKS with fields like TELEMETRY_PERIOD
+### 3.1.6 Syncd
+FlexCounter should be extended to collect PG stats.
+### 3.1.7 SAI
+No update is needed.
+### 3.1.8 gRPC
+No update is needed. Telemetry enabled image starts the service by default, and this should provide
+access to data in counters DB.
```

Comment:
the watermarks need to be integrated in telemetry virtual path, but this can be phase II.